### PR TITLE
Events and Triggers

### PR DESCRIPTION
**Pull request by wthrowe**
### Types of changes:

- [ ] Bugfix
- [X] New feature

### Component:

- [X] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

### Further comments

I've gone with putting the Events and Triggers in the global cache, and as a result they cannot have state.  If we want stateful triggers moving them to the DataBox would not be difficult.

### CODE REVIEW

**Comment by wthrowe** [comment]
Rebased to handle interaction with #485.

File: src/Evolution/EventsAndTriggers/Completion.hpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    auto al_gore =
+        Parallel::get_parallel_component<Component>(cache)[array_index]
+        .ckLocal();
+    al_gore->set_terminate(true);
+  }
+};
+
+/// \cond
+// FIXME move to cpp
+PUP::able::PUP_ID Completion::my_PUP_ID = 0;  // NOLINT
```
Comment (kidder) [inline src/Evolution/EventsAndTriggers/Completion.hpp]: squash this away

File: src/Evolution/EventsAndTriggers/Event.hpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#include "DataStructures/DataBox.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Utilities/FakeVirtual.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \ingroup EventsAndTriggersGroup
+namespace Events {
+class Completion;
+}
```
Comment (kidder) [inline src/Evolution/EventsAndTriggers/Event.hpp]: // namespace Events

File: src/Evolution/EventsAndTriggers/Event.hpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  /// \cond
+  Event() = default;
+  Event(const Event&) = default;
+  Event(Event&&) = default;
+  Event& operator=(const Event&) = default;
+  Event& operator=(Event&&) = default;
+  /// \endcond
+
+ public:
+  ~Event() = default;
```
Comment (kidder) [inline src/Evolution/EventsAndTriggers/Event.hpp]: override

File: src/Evolution/EventsAndTriggers/LogicalTriggers.hpp [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+  void pup(PUP::er& p) {  // NOLINT
+    p | combined_triggers_;
+  }
+
+ private:
+  std::vector<std::unique_ptr<Trigger>> combined_triggers_;
+};
+
+/// \cond
+// FIXME move to cpp
```
Comment (kidder) [inline src/Evolution/EventsAndTriggers/LogicalTriggers.hpp]: ditto

File: src/Evolution/EventsAndTriggers/Trigger.hpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+#include "Utilities/FakeVirtual.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \ingroup EventsAndTriggersGroup
+namespace Triggers {
+class Always;
+class Not;
+class And;
+class Or;
+}
```
Comment (kidder) [inline src/Evolution/EventsAndTriggers/Trigger.hpp]: // namespace

File: src/Evolution/EventsAndTriggers/Trigger.hpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  /// \cond
+  Trigger() = default;
+  Trigger(const Trigger&) = default;
+  Trigger(Trigger&&) = default;
+  Trigger& operator=(const Trigger&) = default;
+  Trigger& operator=(Trigger&&) = default;
+  /// \endcond
+
+ public:
+  ~Trigger() = default;
```
Comment (kidder) [inline src/Evolution/EventsAndTriggers/Trigger.hpp]: override

File: src/Evolution/EventsAndTriggers/Trigger.hpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+  WRAPPED_PUPable_abstract(Trigger);  // NOLINT
+
+  using creatable_classes =
+      tmpl::list<Triggers::Always, Triggers::Not, Triggers::And, Triggers::Or>;
```
Comment (kidder) [inline src/Evolution/EventsAndTriggers/Trigger.hpp]: alphabetize

File: src/Options/Options.hpp [hunk 8]
```diff
@@ -8,14 +8,19 @@
...
 #include <string>
 #include <utility>
 
 #include "ErrorHandling/Error.hpp"
 
-class Option;
+/// \cond
+namespace YAML {
+class Node;
+}
```
Comment (kidder) [inline src/Options/Options.hpp]: // namespace

File: src/Evolution/EventsAndTriggers/Event.hpp [hunk 9]
```diff
@@ -32,7 +39,9 @@ class Event : public PUP::able {
 
   WRAPPED_PUPable_abstract(Event);  // NOLINT
 
-  using creatable_classes = tmpl::list<Events::Completion>;
+  using creatable_classes =
+      tmpl::append<Event_detail::default_events<KnownEvents>,
```
Comment (kidder) [inline src/Evolution/EventsAndTriggers/Event.hpp]: check for duplicates

File: src/Evolution/EventsAndTriggers/Trigger.hpp [hunk 10]
```diff
@@ -35,7 +47,8 @@ class Trigger : public PUP::able {
   WRAPPED_PUPable_abstract(Trigger);  // NOLINT
 
   using creatable_classes =
-      tmpl::list<Triggers::Always, Triggers::Not, Triggers::And, Triggers::Or>;
+      tmpl::append<Trigger_detail::logical_triggers<KnownTriggers>,
```
Comment (kidder) [inline src/Evolution/EventsAndTriggers/Trigger.hpp]: check for duplicates

File: src/Time/Triggers/EveryNSlabs.hpp [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+  using argument_tags = tmpl::list<Tags::TimeId>;
+
+  bool operator()(const TimeId& time_id) const noexcept {
+    const size_t slab_number = time_id.slab_number;
+    return slab_number >= offset_ and (slab_number - offset_) % interval_ == 0;
+  }
+
+  // clang-tidy: google-runtime-references
+  void pup(PUP::er& p) {  // NOLINT
```
Comment (kidder) [inline src/Time/Triggers/EveryNSlabs.hpp]: noexcept

File: tests/Unit/Time/Test_PastTime.cpp [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  Parallel::register_derived_classes_with_charm<Trigger<TimeTriggers>>();
+
+  const auto trigger =
+      test_factory_creation<Trigger<TimeTriggers>>("  PastTime: -7.");
+
+  const auto sent_trigger = serialize_and_deserialize(trigger);
+
+  const Slab slab(-10., 10.);
+
+  {
```
Comment (kidder) [inline tests/Unit/Time/Test_PastTime.cpp]: combine into lambda

**Review by kidder — CHANGES_REQUESTED** [review]
rebase, fix, and squash

File: src/Options/ParseOptions.hpp [hunk 13]
```diff
@@ -590,12 +590,35 @@ struct convert<Options_detail::CreateWrapper<T>> {
 template <typename K, typename V, typename H, typename P>
 struct create_from_yaml<std::unordered_map<K, V, H, P>> {
   static std::unordered_map<K, V, H, P> create(const Option& options) {
-    std::map<K, V> ordered = options.parse_as<std::map<K, V>>();
+    // This shared_ptr stuff is a hack to work around the inability to
+    // extract keys from maps before C++17.  One we require C++17 this
```
Comment (kidder) [inline src/Options/ParseOptions.hpp]: One -> once

File: src/Evolution/EventsAndTriggers/Trigger.hpp [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Utilities/FakeVirtual.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \ingroup EventsAndTriggersGroup
+namespace Triggers {
+template <typename>
```
Comment (kidder) [inline src/Evolution/EventsAndTriggers/Trigger.hpp:14]: alphabetize

File: tests/Unit/Evolution/EventsAndTriggers/Test_EventsAndTriggers.cpp [hunk 15]
```diff
@@ -0,0 +1,146 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+// This file checks the Completion event and the basic logical
+// triggers (Always, Not, And, and Or).
```
Comment (kidder) [inline tests/Unit/Evolution/EventsAndTriggers/Test_EventsAndTriggers.cpp]: alphabetize

File: src/Evolution/EventsAndTriggers/Actions/RunEventsAndTriggers.hpp [hunk 16]
```diff
@@ -0,0 +1,44 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
```
Comment (kidder) [inline src/Evolution/EventsAndTriggers/Actions/RunEventsAndTriggers.hpp:3]: use get from ConstGlobalCache with a base class

File: tests/Unit/Evolution/EventsAndTriggers/Test_EventsAndTriggers.cpp [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+
+void run_events_and_triggers(const EventsAndTriggersType& events_and_triggers,
+                             const bool expected) {
+  // Test pup
+  Parallel::register_derived_classes_with_charm<Event<DefaultClasses>>();
+  Parallel::register_derived_classes_with_charm<Trigger<DefaultClasses>>();
+  ActionTesting::ActionRunner<Metavariables> runner{{
+    serialize_and_deserialize(events_and_triggers)}};
+
+  auto box = db::create<db::AddTags<>>();
```
Comment (kidder) [inline tests/Unit/Evolution/EventsAndTriggers/Test_EventsAndTriggers.cpp]: const

File: src/Time/Triggers/EveryNSlabs.hpp [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    using type = size_t;
+    static constexpr OptionString help{"First slab to trigger on."};
+    static type default_value() { return 0; }
+  };
+
+  using options = tmpl::list<N, Offset>;
+  static constexpr OptionString help{
+    "Trigger every N time slabs after a given offset."};
+
+  explicit EveryNSlabs(const size_t interval, const size_t offset) noexcept
```
Comment (kidder) [inline src/Time/Triggers/EveryNSlabs.hpp]: no explicit

**Review by kidder — CHANGES_REQUESTED** [review]
squash

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**